server/src/main/java/io/druid/segment/realtime/firehose/ReplayableFirehoseFactory.java

```diff
@@ -0,0 +1,313 @@
+  private static final int DEFAULT_MAX_TEMP_FILE_SIZE = 250000000;
+  private static final int DEFAULT_READ_FIREHOSE_RETRIES = 3;
+
+  private final FirehoseFactory delegateFactory;
+  private final boolean reportParseExceptions;
+  private final int maxTempFileSize;
+  private final int readFirehoseRetries;
+
+  private ReplayableFirehose firehose;
+
+  @JsonCreator
+  public ReplayableFirehoseFactory(
+      @JsonProperty("delegate") FirehoseFactory delegateFactory,
+      @JsonProperty("reportParseExceptions") Boolean reportParseExceptions,
+      @JsonProperty("maxTempFileSize") Integer maxTempFileSize,
+      @JsonProperty("readFirehoseRetries") Integer readFirehoseRetries
+  )
+  {
+    Preconditions.checkNotNull(delegateFactory, "delegate cannot be null");
+    Preconditions.checkArgument(
+        !(delegateFactory instanceof ReplayableFirehoseFactory),
+        "Refusing to wrap another ReplayableFirehoseFactory"
+    );
+
+    this.delegateFactory = delegateFactory;
+    this.reportParseExceptions = reportParseExceptions == null
+                                 ? DEFAULT_REPORT_PARSE_EXCEPTIONS
+                                 : reportParseExceptions;
+    this.maxTempFileSize = maxTempFileSize == null ? DEFAULT_MAX_TEMP_FILE_SIZE : maxTempFileSize;
+    this.readFirehoseRetries = readFirehoseRetries == null ? DEFAULT_READ_FIREHOSE_RETRIES : readFirehoseRetries;
+
+    log.info(this.toString());
+  }
+
+  @Override
+  public Firehose connect(InputRowParser parser) throws IOException
+  {
+    if (firehose == null) {
+      firehose = new ReplayableFirehose(parser);
+    } else {
+      log.info("Rewinding and returning existing firehose");
+      firehose.rewind();
+    }
+
+    return firehose;
+  }
+
+  @Override
+  public void close() throws IOException
+  {
+    if (firehose != null) {
+      firehose.deleteTempFiles();
+      firehose.close();
+    }
+
+    firehose = null;
+  }
+
+  public class ReplayableFirehose implements Firehose
+  {
+    private final JsonFactory factory = new DefaultObjectMapper(
+        new SmileFactory().enable(SmileGenerator.Feature.CHECK_SHARED_STRING_VALUES)
+    ).getFactory();
+
+    private final List<File> files = new ArrayList<>();
+    private final List<String> dimensions;
+
+    private int fileIndex = 0;
+    private JsonParser jsonParser;
+    private Iterator<Row> it;
+
+    public ReplayableFirehose(InputRowParser parser) throws IOException
+    {
+      long counter = 0, totalBytes = 0, unparseable = 0, retryCount = 0;
+      Set<String> dimensionScratch = new HashSet<>();
+
+      File tmpDir = Files.createTempDir();
+      tmpDir.deleteOnExit();
+
+      long startTime = System.nanoTime();
+      do {
+        deleteTempFiles();
+        try (Firehose delegateFirehose = delegateFactory.connect(parser)) {
+          while (delegateFirehose.hasMore()) {
+            File tmpFile = File.createTempFile("replayable-", null, tmpDir);
+            tmpFile.deleteOnExit();
```

Comment:
I believe tmpDir.deleteOnExit() will fail if there are still files in the directory; for reference: http://docs.oracle.com/javase/7/docs/api/java/io/File.html#delete()